enhancements/network/core-network-policies.md

```diff
@@ -0,0 +1,532 @@
+Admin Network Policy can be used to override our policies until a fix
+can be delivered in a z-stream.
+
+
+#### Policies Too Open
+
+The converse can also happen… we may write a policy that a customer
+finds is too open, or they may wish to impose additional restrictions
+that we may allow in general (for instance, something could allow a
+webhook to be called off-cluster, but they may want to forbid all
+off-cluster traffic from the namespace since they do not use that
+feature).
+
+Customers can write admin network policies that override the network
+policies that we ship.
+
+
+#### Challenges Debugging and Supporting
+
+When a network policy blocks traffic, it may be blocking traffic of an
+internal service communication causing a mysterious failure message to
+the user.
+
+You can use the Network Observability tool to audit network policy
+blocks.  There should not be any in the OpenShift namespaces.
+
+We also need to provide this information to the support organization
+in a must-gather.  We will augment the must-gather tool to include
+information about network policy blocks in OpenShift namespaces.
+
+
+### Drawbacks
+
+None foreseen.
+
+
+## Test Plan
+
+- The existing e2e and integration tests should catch network policy errors where they are too restrictive
+
+- We will need to add new tests to ensure that all “openshift-” namespaces have the default deny policy defined.
+
+- We should add a test that runs alongside the existing e2e tests to
+  monitor for connections being blocked by network policy in openshift
+  namespaces.  There should be no such failures in normal operation.
+
+- We need to have tests for the new functionality in the admission controller and for the cluster network operator
+
+
+## Graduation Criteria
+
+GA:
+
+- Cluster Network Operator changes complete:
+- Namespace Admission controller change complete
+- Must-gather able to get all recent network policy connection blocks in “openshift-” namespaces
+- Customers must be able to see network policy connection blocks using the observability tools
+- The test changes as described above must be complete
+- Documentation will be updated to describe how to:
+    - Override our network policies using admin network policy
+    - Apply a default-deny rule to all OpenShift namespaces
+    - Apply a default-deny rule to all cluster namespaces
+    - Disable our policies in a KCS
+
+
+## Upgrade / Downgrade Strategy
+
+
+#### Upgrades:
+
+
+
+1. The new API server will roll out and will start labeling namespaces
+
+2. The Cluster Network Operator will be upgraded and upon restart it
+will scan all namespaces and fix the labels to handle namespaces that
+have not changed
+
+Since nothing needs to use the labels immediately, there is no concern about ordering and racing, as long as the change happens during the upgrade.
+
+At which point, the cluster administrator can use the new labels to easily apply default policies to all of the OpenShift namespaces.	
```

Comment:
When do the new policies get created? How do you ensure there won't be problems when some namespaces have policies and some don't, and some have labels and some don't?

In my earlier proposal (#613) I suggested we'd have to roll this out over several releases to avoid problems...

Reply:
This will likely take a while to roll out since each team will need to add policy to their namespaces.  And I suspect they will need assistance developing the policy, which is where ACS may be able to help for static analysis of objects and dynamic analysis of trafic and it will suggest policies based on that.